### PR TITLE
ISSUE-10 distributed and distributed ldap deployment test

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]
Looks good to me so far. How long does iit take to run these tests?

**Comment by Arsnael** [comment]
> Looks good to me so far. How long does iit take to run these tests?

Just one test... But with more it can take quite long, as we restart all containers between each test.

**Comment by chibenwa** [comment]
I think we can live with just CLI for LDAP but for distributed we need IMAP SMTP and JMAP as well IMO

> Just one test for now... But with more it can take quite long, as we restart all containers between each test.

How much time for a test?

(Could we try to parallelize containers startups?)

File: openpaas-james/deployment-tests/distributed/src/test/java/com/linagora/openpaas/deployment/OpenpaasJamesDistributedExtension.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            .waitingFor(Wait.forLogMessage(".*Web admin server started.*\\n", ONE_TIME));
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext extensionContext) {
+        cassandra.start();
+        elasticsearch.start();
+        rabbitmq.start();
+        s3.start();
+        distributed.start();
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/distributed/src/test/java/com/linagora/openpaas/deployment/OpenpaasJamesDistributedExtension.java:81]:
```
Runnables.runParralell(
    cassandra::start,
    elasticSearch::start,
    rabbitmq::start,
    s3::start);
distributed.start();
```

File: openpaas-james/deployment-tests/distributed/src/test/java/com/linagora/openpaas/deployment/OpenpaasJamesDistributedExtension.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    @Override
+    public void afterEach(ExtensionContext extensionContext) {
+        cassandra.stop();
+        elasticsearch.stop();
+        rabbitmq.stop();
+        s3.stop();
+        distributed.stop();
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/distributed/src/test/java/com/linagora/openpaas/deployment/OpenpaasJamesDistributedExtension.java:90]: Idem use `Runnables.runParralell`

**Comment by Arsnael** [comment]
> I think we can live with just CLI for LDAP but for distributed we need IMAP SMTP and JMAP as well IMO

I never said I would not do it? It's a draft :) I just posted yesterday the advancement of my work. 

I will care about the time the moment everything works...

IMAP SMTP and JMAP seem working, except that I have something weird with IMAP and SMTP tests. Like it's 3 tests in that class. If I run them one by one seems fine, but all together seems not stable, don't know yet...

**Comment by chibenwa** [comment]
> If I run them one by one seems fine, but all together seems not stable, don't know yet...

Likely DB persistance, I would suggest enforcing different users are used for each test.

**Comment by Arsnael** [comment]
Seems more like concurrency to me, as when checking the code when we do a container.stop() we do remove the container

File: openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java [hunk 3]
```diff
@@ -25,6 +25,8 @@
     @Override
     @BeforeEach
     public void setUp() throws Exception {
+        Thread.sleep(2000);
```
Comment (Arsnael) [inline openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java:28]: At the end I had to put a sleep of 2s in every BeforeEach of tests, as I noticed we seem to have a little latency between the log we are waiting for assuming james is ready, and when the server is really ready. With those, the tests are not unstable anymore.

It's ugly I know...

File: openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/CliContract.java [hunk 4]
```diff
@@ -12,6 +12,8 @@
 
     @Test
     default void cliShouldWork() throws Exception {
+        Thread.sleep(2000);
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/CliContract.java:15]: While being acceptable as a (very) short term solution, I would suggest putting an INFO log in `JMXServer.java` and awaiting that log here.

Can you contribute the log addition upstream in apache/james-project , and work asynchronously on removing this thread.sleep?

File: openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java (hunk 3 above)
Comment (chibenwa) [reply openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java:28]: While being acceptable as a (very) short term solution, I would suggest putting an INFO log in `JMXServer.java` and awaiting that log here.

Can you contribute the log addition upstream in apache/james-project , and work asynchronously on removing this thread.sleep?

File: openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    @SuppressWarnings("resource")
+    private GenericContainer<?> createLdap() {
+        return new GenericContainer<>("dinkel/openldap:latest")
+            .withNetworkAliases("ldap")
+            .withNetwork(network)
+            .withEnv("SLAPD_DOMAIN", "james.org")
+            .withEnv("SLAPD_PASSWORD", "mysecretpassword")
+            .withEnv("SLAPD_CONFIG_PASSWORD", "mysecretpassword")
+            .withClasspathResourceMapping("populate.ldif", "/etc/ldap/prepopulate/populate.ldif", BindMode.READ_ONLY);
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java:79]: Let's see if this passes on the (future) CI but withClasspathResourceMapping was not working for similar use cases.

We ended up building a custom image on the fly instead.

File: openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public void beforeEach(ExtensionContext extensionContext) {
+        cassandra.start();
+        elasticsearch.start();
+        rabbitmq.start();
+        s3.start();
+        ldap.start();
+        distributedLdap.start();
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java:97]: parallelize this?

File: openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+public class OpenpaasJamesDistributedLdapExtension implements BeforeEachCallback, AfterEachCallback {
+    private static final int ONE_TIME = 1;
+
+    private final Network network;
+    private final GenericContainer<?> cassandra;
+    private final GenericContainer<?> elasticsearch;
+    private final GenericContainer<?> rabbitmq;
+    private final GenericContainer<?> s3;
+    private final GenericContainer<?> distributedLdap;
```
Comment (chibenwa) [inline openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java:22]: distributedLdap -> james

File: openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java (hunk 5 above)
Comment (Arsnael) [reply openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java:79]: This works locally, I could not find any other way.

File: openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java [hunk 8]
```diff
@@ -14,9 +14,9 @@
 import com.google.inject.Injector;
 
 public abstract class ImapAndSmtpContract extends DeploymentValidation {
-    abstract ExternalJamesConfiguration configuration();
+    protected abstract ExternalJamesConfiguration configuration();
 
-    abstract GenericContainer<?> container();
+    protected abstract GenericContainer<?> container();
```
Comment (Arsnael) [inline openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java:19]: somehow I could not extend `DistributedLdapImapAndSmtpTest` with `ImapAndSmtpContract` without making the abstract methods protected... The compilation would just keep failing, while other modules were ok on that... Could not figure out why honestly...

File: openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java (hunk 5 above)
Comment (chibenwa) [reply openpaas-james/deployment-tests/distributed-ldap/src/test/java/deployment/OpenpaasJamesDistributedLdapExtension.java:79]: Fine for now IMO

**Comment by Arsnael** [comment]
It's not the prettiest work but the deployment tests are green finally :)

**Review by LanKhuat — APPROVED** [review]

File: openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java (hunk 3 above)
Comment (Arsnael) [reply openpaas-james/deployment-tests/common/src/main/java/com/linagora/openpaas/deployment/ImapAndSmtpContract.java:28]: https://github.com/linagora/james-project/pull/4234

**Pull request merged**